bpf/bpf_lxc.c

```diff
@@ -743,7 +743,27 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_ARP)
 int tail_handle_arp(struct __ctx_buff *ctx)
 {
 	union macaddr mac = NODE_MAC;
-	return arp_respond(ctx, &mac, 0);
+	union macaddr smac;
+	__be32 sip;
+	__be32 tip;
+
+	/* Pass any unknown ARP requests to the Linux stack */
+	if (!arp_validate(ctx, &mac, &smac, &sip, &tip))
+		return CTX_ACT_OK;
+
+	/*
+	 * The endpoint is expected to make ARP requests for its gateway IP.
+	 * Most of the time, the gateway IP configured on the endpoint is
+	 * IPV4_GATEWAY but it may not be the case if after cilium agent reload
+	 * a different gateway is chosen. In such a case, existing endpoints
+	 * will have an old gateway configured. Since we don't know the IP of
+	 * previous gateways, we answer requests for all IPs with the exception
+	 * of the LXC IP.
```

Comment:
This comment is detailed and provides a lot of context, but it misses the background of why we don't respond to ARP requests for the endpoint's own IP---that is to say, to avoid breaking duplicate address detection.

Reply:
Added.